### PR TITLE
raftstore: purge raft engine periodically

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What problem does this PR solve?

`RaftEngine` has a method `purge_expired_files`. Now in TiKV the function won't be called. It's ok for the rocksdb implementation, but for other implementations it could be necessary.

### What is changed and how it works?

Call `purge_expired_files` every 10 minutes.

### Check List 

Tests 
- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 
* purge raft engine periodically

### CODE REVIEW

File: components/raftstore/src/store/config.rs [hunk 1]
```diff
@@ -66,6 +66,8 @@ pub struct Config {
     // When the approximate size of raft log entries exceed this value,
     // gc will be forced trigger.
     pub raft_log_gc_size_limit: ReadableSize,
+    // Old logs in Raft engine needs to be purged peridically.
+    pub raft_engine_purge_interval: ReadableDuration,
```
Comment (Little-Wallace) [inline components/raftstore/src/store/config.rs:70]: There has been a configure `raft_log_gc_tick_interval` used to check whether the log is too large. Could we use it to check old log of `raft-engine`?

Comment (hicqu) [reply components/raftstore/src/store/config.rs:70]: They are different. `raft_log_gc_tick_interval` is for one region but the latter is for whole engine.

**Review by Little-Wallace — APPROVED** [review]
LGTM

File: components/raftstore/src/store/config.rs [hunk 2]
```diff
@@ -206,6 +208,7 @@ impl Default for Config {
             // Assume the average size of entries is 1k.
             raft_log_gc_count_limit: split_size * 3 / 4 / ReadableSize::kb(1),
             raft_log_gc_size_limit: split_size * 3 / 4,
+            raft_engine_purge_interval: ReadableDuration::minutes(10),
```
Comment (zhangjinpeng1987) [inline components/raftstore/src/store/config.rs:211]: It is too long, 10 seconds is proper.

File: components/raftstore/src/store/fsm/store.rs [hunk 3]
```diff
@@ -2344,6 +2346,23 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: Transport, C: PdClient>
...
     }
+
+    fn register_raft_engine_purge_tick(&self) {
+        self.ctx.schedule_store_tick(
+            StoreTick::RaftEnginePurge,
+            self.ctx.cfg.raft_engine_purge_interval.0,
+        )
+    }
+
+    fn on_raft_engine_purge_tick(&self) {
```
Comment (zhangjinpeng1987) [inline components/raftstore/src/store/fsm/store.rs:2357]: Please move the calculation out of the raftstore thread.

File: components/raftstore/src/store/fsm/peer.rs [hunk 4]
```diff
@@ -3215,40 +3219,31 @@ where
...
-        } else {
-            compact_idx = replicated_idx;
-        }
 
-        // Have no idea why subtract 1 here, but original code did this by magic.
-        assert!(compact_idx > 0);
-        compact_idx -= 1;
-        if compact_idx < first_idx {
-            // In case compact_idx == first_idx before subtraction.
+        if replicated_idx < first_idx {
```
Comment (zhangjinpeng1987) [inline components/raftstore/src/store/fsm/peer.rs:3223]: How about force_compact = true.

File: components/raftstore/src/store/fsm/peer.rs [hunk 5]
```diff
@@ -3215,37 +3219,32 @@ where
...
-        }
 
-        // Have no idea why subtract 1 here, but original code did this by magic.
-        assert!(compact_idx > 0);
-        compact_idx -= 1;
-        if compact_idx < first_idx {
-            // In case compact_idx == first_idx before subtraction.
+        if !force_compact
+            && (replicated_idx < first_idx
+                || replicated_idx - first_idx <= self.ctx.cfg.raft_log_gc_threshold)
```
Comment (hicqu) [inline components/raftstore/src/store/fsm/peer.rs:3225]: @BusyJay please take a look here, Can `raft_log_gc_threshold` be reduced to `16` or a less value?

File: components/raftstore/src/store/fsm/peer.rs [hunk 6]
```diff
@@ -3214,40 +3222,42 @@ where
...
-            || replicated_idx - first_idx <= self.ctx.cfg.raft_log_gc_threshold
-        {
-            return;
-        } else {
-            compact_idx = replicated_idx;
-        }
 
-        // Have no idea why subtract 1 here, but original code did this by magic.
-        assert!(compact_idx > 0);
-        compact_idx -= 1;
```
Comment (hicqu) [inline components/raftstore/src/store/fsm/peer.rs:3235]: I removed `compact_idx -= 1`, PTAL @BusyJay

Comment (zhangjinpeng1987) [inline components/raftstore/src/store/fsm/peer.rs:3235]: Let's keep this behavior, or get to know why subtract 1 before?

File: components/raftstore/src/store/fsm/peer.rs [hunk 7]
```diff
@@ -109,6 +109,8 @@ where
     pub receiver: Receiver<PeerMsg<EK>>,
     /// when snapshot is generating or sending, skip split check at most REGION_SPLIT_SKIT_MAX_COUNT times.
     skip_split_count: usize,
+    /// Sometimes applied raft logs won't be compacted in time, don't keep them too long time.
```
Comment (zhangjinpeng1987) [inline components/raftstore/src/store/fsm/peer.rs:112]: The comment is not very clear.

**Review by zhangjinpeng1987 — APPROVED** [review]
LGTM

File: components/raftstore/src/store/fsm/peer.rs [hunk 8]
```diff
@@ -3214,40 +3224,49 @@ where
...
-            || replicated_idx - first_idx <= self.ctx.cfg.raft_log_gc_threshold
-        {
+
+        if !force_compact && replicated_idx < first_idx {
             return;
-        } else {
-            compact_idx = replicated_idx;
         }
 
+        if !force_compact && replicated_idx - first_idx < self.ctx.cfg.raft_log_gc_threshold {
```
Comment (zhangjinpeng1987) [inline components/raftstore/src/store/fsm/peer.rs:3234]: If there is only one raft entry left that is COMPACT_LOG type, also will trigger raft log gc after 6 rounds checking.

File: components/raftstore/src/store/fsm/peer.rs [hunk 9]
```diff
@@ -3223,40 +3233,57 @@ where
...
+            // In the current implementation one compaction can't delete all stale Raft logs.
+            // There will be at least 3 entries left after one compaction:
+            // |------------- entries needs to be compacted ----------|
+            // [entries...][the entry at `compact_idx`][the last entry][new compaction entry]
+            //             |-------------------- entries will be left ----------------------|
+            if last_idx - first_idx < 3 {
+                return;
+            }
+            self.fsm.skip_gc_raft_log_ticks += 1;
+            if self.fsm.skip_gc_raft_log_ticks <= self.ctx.cfg.raft_log_reserve_max_ticks {
```
Comment (zhangjinpeng1987) [inline components/raftstore/src/store/fsm/peer.rs:3253]: s/</<=

**Review by zhangjinpeng1987 — APPROVED** [review]
LGTM

**Pull request merged**